### ISSUE TITLE
Validation results from train.py and val.py differ

### ISSUE THREAD

**Issue by rgaiacs** [ultralytics/yolov5#8552]
### Search before asking

- [X] I have searched the YOLOv5 [issues](https://github.com/ultralytics/yolov5/issues) and found no similar bug report.

### YOLOv5 Component

Validation

### Bug

I executed

```
$ yaml=first-carpal-size-1-fold-0.yaml
$ project=debug
$ epoch=32
$ python train.py \
       --img 640 \
       --batch 16 \
       --epochs $epoch \
       --data $yaml \
       --weights yolov5s.pt \
       --project $project \
       --name $epoch
$ python val.py \
       --img 640 \
       --batch 16 \
       --data $yaml \
       --weights $project/32/weights/best.pt \
       --task val \
       --save-txt \
       --save-hybrid \
       --save-conf \
       --save-json \
       --project $project \
       --name val
```

The training executed as expected.

```
train: weights=yolov5s.pt, cfg=, data=first-carpal-size-1-fold-0.yaml, hyp=data/hyps/hyp.scratch-low.yaml, epochs=32, batch_size=16, imgsz=640, rect=False, resume=False, nosave=False, noval=False, noautoanchor=False, noplots=\
False, evolve=None, bucket=, cache=None, image_weights=False, device=, multi_scale=False, single_cls=False, optimizer=SGD, sync_bn=False, workers=8, project=debug, name=32, exist_ok=False, quad=False, cos_lr=False, label_smoo\
thing=0.0, patience=100, freeze=[0], save_period=-1, local_rank=-1, entity=None, upload_dataset=False, bbox_interval=-1, artifact_alias=latest
github: ⚠️ YOLOv5 is out of date by 40 commits. Use `git pull` or `git clone https://github.com/ultralytics/yolov5` to update.
YOLOv5 🚀 2022-6-14 Python-3.10.4 torch-1.11.0 CUDA:0 (Quadro RTX 4000, 7960MiB)

hyperparameters: lr0=0.01, lrf=0.01, momentum=0.937, weight_decay=0.0005, warmup_epochs=3.0, warmup_momentum=0.8, warmup_bias_lr=0.1, box=0.05, cls=0.5, cls_pw=1.0, obj=1.0, obj_pw=1.0, iou_t=0.2, anchor_t=4.0, fl_gamma=0.0, \
hsv_h=0.015, hsv_s=0.7, hsv_v=0.4, degrees=0.0, translate=0.1, scale=0.5, shear=0.0, perspective=0.0, flipud=0.0, fliplr=0.5, mosaic=1.0, mixup=0.0, copy_paste=0.0
Weights & Biases: run 'pip install wandb' to automatically track and visualize YOLOv5 🚀 runs (RECOMMENDED)
TensorBoard: Start with 'tensorboard --logdir debug', view at http://localhost:6006/
Overriding model.yaml nc=80 with nc=6

                 from  n    params  module                                  arguments
  0                -1  1      3520  models.common.Conv                      [3, 32, 6, 2, 2]
  1                -1  1     18560  models.common.Conv                      [32, 64, 3, 2]
  2                -1  1     18816  models.common.C3                        [64, 64, 1]
  3                -1  1     73984  models.common.Conv                      [64, 128, 3, 2]
  4                -1  2    115712  models.common.C3                        [128, 128, 2]
  5                -1  1    295424  models.common.Conv                      [128, 256, 3, 2]
  6                -1  3    625152  models.common.C3                        [256, 256, 3]
  7                -1  1   1180672  models.common.Conv                      [256, 512, 3, 2]
  8                -1  1   1182720  models.common.C3                        [512, 512, 1]
  9                -1  1    656896  models.common.SPPF                      [512, 512, 5]
 10                -1  1    131584  models.common.Conv                      [512, 256, 1, 1]
 11                -1  1         0  torch.nn.modules.upsampling.Upsample    [None, 2, 'nearest']
 12           [-1, 6]  1         0  models.common.Concat                    [1]
 13                -1  1    361984  models.common.C3                        [512, 256, 1, False]
 14                -1  1     33024  models.common.Conv                      [256, 128, 1, 1]
 15                -1  1         0  torch.nn.modules.upsampling.Upsample    [None, 2, 'nearest']
 16           [-1, 4]  1         0  models.common.Concat                    [1]
 17                -1  1     90880  models.common.C3                        [256, 128, 1, False]
 18                -1  1    147712  models.common.Conv                      [128, 128, 3, 2]
 19          [-1, 14]  1         0  models.common.Concat                    [1]
 20                -1  1    296448  models.common.C3                        [256, 256, 1, False]
 21                -1  1    590336  models.common.Conv                      [256, 256, 3, 2]
 22          [-1, 10]  1         0  models.common.Concat                    [1]
 23                -1  1   1182720  models.common.C3                        [512, 512, 1, False]
 24      [17, 20, 23]  1     29667  models.yolo.Detect                      [6, [[10, 13, 16, 30, 33, 23], [30, 61, 62, 45, 59, 119], [116, 90, 156, 198, 373, 326]], [128, 256, 512]]
Model summary: 270 layers, 7035811 parameters, 7035811 gradients

Transferred 343/349 items from yolov5s.pt
AMP: checks passed ✅
Scaled weight_decay = 0.0005
optimizer: SGD with parameter groups 57 weight (no decay), 60 weight, 60 bias
albumentations: Blur(always_apply=False, p=0.01, blur_limit=(3, 7)), MedianBlur(always_apply=False, p=0.01, blur_limit=(3, 7)), ToGray(always_apply=False, p=0.01), CLAHE(always_apply=False, p=0.01, clip_limit=(1, 4.0), tile_g\
rid_size=(8, 8))
train: Scanning '/home/raniere/Documents/yolov5-carpal/datasets/First Carpal Bone (DMPLO) Detection YOLO n=1 0/labels/train.cache' images and labels... 16 found, 0 missing, 0 empty, 0 corrupt: 100%|██████████| 16/16 [00:00<?,
val: Scanning '/home/raniere/Documents/yolov5-carpal/datasets/First Carpal Bone (DMPLO) Detection YOLO n=1 0/labels/val.cache' images and labels... 4 found, 0 missing, 0 empty, 0 corrupt: 100%|██████████| 4/4 [00:00<?, ?it/s]
Plotting labels to debug/32/labels.jpg...

AutoAnchor: 4.75 anchors/target, 1.000 Best Possible Recall (BPR). Current anchors are a good fit to dataset ✅
Image sizes 640 train, 640 val
Using 8 dataloader workers
Logging results to debug/32
Starting training for 32 epochs...

[...]

32 epochs completed in 0.011 hours.
Optimizer stripped from debug/32/weights/last.pt, 14.4MB
Optimizer stripped from debug/32/weights/best.pt, 14.4MB

Validating debug/32/weights/best.pt...
Fusing layers...
Model summary: 213 layers, 7026307 parameters, 0 gradients
               Class     Images     Labels          P          R     mAP@.5 mAP@.5:.95: 100%|██████████| 1/1 [00:00<00:00,  6.45it/s]
                 all          4         22    0.00535      0.292     0.0887     0.0118
              radius          4          4    0.00664        0.5    0.00758    0.00122
 proximal carpal row          4          4    0.00806        0.5     0.0215     0.0028
   distal carpal row          4          4     0.0036       0.25    0.00447    0.00134
     acessory carpal          4          4          0          0          0          0
          metacarpal          4          4     0.0138        0.5      0.499     0.0657
   first carpal bone          4          2          0          0          0          0
Warning: Plotting error for debug/32/results.csv: index 1 is out of bounds for axis 1 with size 1
No artists with labels found to put in legend.  Note that artists whose label start with an underscore are ignored when legend() is called with no argument.
Results saved to debug/32
```

I'm aware that I trained using very little images. Now, the validation is repeated:

```
val: data=/home/raniere/Documents/yolov5-carpal/yolov5/data/first-carpal-size-1-fold-0.yaml, weights=['debug/32/weights/best.pt'], batch_size=16, imgsz=640, conf_thres=0.001, iou_thres=0.6, task=val, device=, workers=8, singl\
e_cls=False, augment=False, verbose=False, save_txt=True, save_hybrid=True, save_conf=True, save_json=True, project=debug, name=val, exist_ok=False, half=False, dnn=False
YOLOv5 🚀 2022-6-14 Python-3.10.4 torch-1.11.0 CUDA:0 (Quadro RTX 4000, 7960MiB)

Fusing layers...
Model summary: 213 layers, 7026307 parameters, 0 gradients
val: Scanning '/home/raniere/Documents/yolov5-carpal/datasets/First Carpal Bone (DMPLO) Detection YOLO n=1 0/labels/val.cache' images and labels... 4 found, 0 missing, 0 empty, 0 corrupt: 100%|██████████| 4/4 [00:00<?, ?it/s]
               Class     Images     Labels          P          R     mAP@.5 mAP@.5:.95: 100%|██████████| 1/1 [00:00<00:00,  1.26it/s]
                 all          4         22          1          1      0.995      0.995
              radius          4          4          1          1      0.995      0.995
 proximal carpal row          4          4          1          1      0.995      0.995
   distal carpal row          4          4          1          1      0.995      0.995
     acessory carpal          4          4          1          1      0.995      0.995
          metacarpal          4          4          1          1      0.995      0.995
   first carpal bone          4          2          1          1      0.995      0.995
Speed: 0.2ms pre-process, 5.9ms inference, 8.1ms NMS per image at shape (16, 3, 640, 640)

Evaluating pycocotools mAP... saving debug/val/best_predictions.json...
loading annotations into memory...
pycocotools unable to run: [Errno 2] No such file or directory: '../datasets/First Carpal Bone (DMPLO) Detection YOLO n=1 0/annotations/instances_val2017.json'
Results saved to debug/val
```

Note that when I run the validation for the second time, all the metrics are different.

### Environment

- YOLOv5 v6.1
- PyTorch 1.11.0 with Python y3.10, CUDA 11.3 and cuDNN 8.2.0
- OS: Ubuntu 22.04

### Are you willing to submit a PR?

- [ ] Yes I'd like to help by submitting a PR!

**Comment by glenn-jocher** [comment]
@rgaiacs --save-hybrid adds predictions to your labels and computes mAP on the combined result after NMS, so you should expect near 1.0 mAP. If you want a correct mAP don't use --save-hybrid.

**Comment by rgaiacs** [comment]
@glenn-jocher Thanks for the reply. Dropping [`--save-hybrid`](https://github.com/ultralytics/yolov5/discussions/2379), the result is what I expected:

```
val: data=/home/raniere/Documents/yolov5-carpal/yolov5/data/first-carpal-size-1-fold-0.yaml, weights=['debug/32/weights/best.pt'], batch_size=16, imgsz=640, conf_thres=0.001, iou_thres=0.6, task=val, device=, workers=8, singl\
e_cls=False, augment=False, verbose=False, save_txt=True, save_hybrid=False, save_conf=True, save_json=True, project=debug, name=val, exist_ok=False, half=False, dnn=False
YOLOv5 🚀 2022-6-14 Python-3.10.4 torch-1.11.0 CUDA:0 (Quadro RTX 4000, 7960MiB)

Fusing layers...
Model summary: 213 layers, 7026307 parameters, 0 gradients
val: Scanning '/home/raniere/Documents/yolov5-carpal/datasets/First Carpal Bone (DMPLO) Detection YOLO n=1 0/la
               Class     Images     Labels          P          R     mAP@.5 mAP@.5:.95: 100%|██████████| 1/1 [0
                 all          4         22    0.00532      0.292     0.0886     0.0118
              radius          4          4    0.00651        0.5    0.00758    0.00122
 proximal carpal row          4          4    0.00813        0.5     0.0208    0.00274
   distal carpal row          4          4    0.00368       0.25     0.0045    0.00135
     acessory carpal          4          4          0          0          0          0
          metacarpal          4          4     0.0136        0.5      0.498     0.0654
   first carpal bone          4          2          0          0          0          0
Speed: 0.2ms pre-process, 6.4ms inference, 7.1ms NMS per image at shape (16, 3, 640, 640)

Evaluating pycocotools mAP... saving debug/val2/best_predictions.json...
loading annotations into memory...
pycocotools unable to run: [Errno 2] No such file or directory: '../datasets/First Carpal Bone (DMPLO) Detection YOLO n=1 0/annotations/instances_val2017.json'
Results saved to debug/val2
4 labels saved to debug/val2/labels
```

**Comment by glenn-jocher** [comment]
@rgaiacs great!
